third_party/move/move-model/src/metadata.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::bail;
+use move_command_line_common::env::read_bool_env_var;
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use std::{
+    fmt,
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
+
+const UNSTABLE_MARKER: &str = "-unstable";
+
+/// Metadata about a compilation result. To maintain serialization
+/// compatibility, this uses a free-form string to represent compiler version
+/// and language version.
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct CompilationMetadata {
+    /// The version of the compiler, as a string.
+    pub compiler_version: String,
+    /// The version of the language, as a string.
+    pub language_version: String,
+    /// A flag indicating whether, at time of creation, the compilation
+    /// result was considered as unstable. Unstable code may have restrictions
+    /// for deployment on production networks.
+    pub unstable: bool,
+}
+
+impl CompilationMetadata {
+    pub fn new(compiler_version: CompilerVersion, language_version: LanguageVersion) -> Self {
+        Self {
+            compiler_version: compiler_version.to_string(),
+            language_version: language_version.to_string(),
+            unstable: compiler_version.unstable() || language_version.unstable(),
+        }
+    }
+
+    pub fn compiler_version(&self) -> anyhow::Result<CompilerVersion> {
+        CompilerVersion::from_str(&self.compiler_version)
+    }
+
+    pub fn language_version(&self) -> anyhow::Result<LanguageVersion> {
+        LanguageVersion::from_str(&self.language_version)
+    }
+
+    /// Returns true of the compilation was created as unstable.
+    pub fn created_as_unstable(&self) -> bool {
+        self.unstable
+    }
+}
+
+/// Represents a compiler version
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, Eq, PartialEq, PartialOrd, Ord)]
+pub enum CompilerVersion {
```

Comment:
Should we do the Semantic versioning here? Maybe ok to have every shingle Major/Minor/Patch - so:
V1_0_0
V2_0_0_Alpha_1 ... etc

Reply:
Alternatively we can call this enum CompilerVersionMajor